// @jmfe/jm-webview/src/basic.js
import { serialize, assign, isAndroid, isIOS } from '@jmfe/jm-common';
import { nativeCall, buildSchema } from './call';

/**
 * 获取产品详情 schema
 * @param {number} id
 * @param {object} [extra]
 * @return {string}
 */
export function getProductSchema(id, extra) {
	return buildSchema('virtual', assign({ 
		category: 'jump', 
		des: 'productDetail',
		skuId: id
	}, extra));
}

/**
 * 进入产品详情页
 * @param {number} id
 * @param {object} [extra]
 */
export function toProduct(id, extra) {
	window.location.href = getProductSchema(id, extra);	
}

/**
 * 获取店铺详情页 schema
 * @param {number} id
 * @param {object} [extra]
 * @return {string}
 */
export function getShopSchema(id, extra) {
	return buildSchema('virtual', assign({
		category: 'jump', 
		des: 'jshopMain',
		shopId: id
	}, extra));
}

/**
 * 进入店铺详情页
 * @param {number} id
 * @param {object} [extra]
 */
export function toShop(id, extra) {
	window.location.href = getShopSchema(id, extra);
}

export function getHomeSchema() {
	if (isAndroid()) return buildSchema('360buy', { type: 1 });
	else if (isIOS()) return buildSchema('virtual', { category: 'jump', des: 'HomePage' }); 
}

/**
 * 回到首页
 */
export function toHome() {
	window.location.href = getHomeSchema();
}

export function getCouponCenterSchema(extra) {
	return buildSchema('virtual',  assign({
		category: 'jump', 
		des: 'couponCenter'
	}, extra));
}

/**
 * 进入领券中心
 * @param {object} [extra]
 */
export function toCouponCenter(extra) {
	window.location.href = getCouponCenterSchema(extra);
}

export function getSecKillSchema(extra) {
	return buildSchema('virtual',  assign({
		category: 'jump', 
		des: 'seckill'
	}, extra));
}

/**
 * 进入秒杀频道
 * @param {object} [extra]
 */
export function toSecKill(extra) {
	window.location.href = getSecKillSchema(extra);
}

export function toFlashSale(extra) {
	console.warn('toFlashSale 已经改名为 toSecKill， 请使用新方法。');
	toSecKill(extra);
}

export function getOrderListSchema(extra) {
	return buildSchema('virtual',  assign({
		category: 'jump', 
		des: 'orderlist'
	}, extra));
}

/**
 * 进入订单列表页
 * @param {number} id
 * @param {object} [extra]
 */
export function toOrderList(extra) {
	window.location.href = getOrderListSchema(extra);
}

/**
 * 跳转到登录页
 * @param {string} [returnUrl]
 */
export function toLogin(returnUrl) {
	returnUrl = returnUrl || window.location.href;
	const url = 'https://passport.m.jd.com/user/login.action';
	let params = { 
		v: 1,
		returnurl: encodeURIComponent(returnUrl)
	};
	window.location.href = url + '?' + serialize(params);
}

/**
 * 打开新的 webview 展示页面
 * @param  {string} url
 */
export function openNewWebView(url) {
	nativeCall('virtual', {
		category: 'jump',
		action: 'to',
		des: 'm',
		url: url
	});
}
