// @jmfe/jm-webview/src/cart.js
import { assign, isAndroid, isIOS, getAppVersion, versionCompare } from '@jmfe/jm-common';
import { nativeCall, buildSchema } from './call';

export function getCartSchema(extra) {
	return buildSchema('virtual', assign({
		category: 'jump', 
		des: 'cartB'
	}, extra));
}

/**
 * 进入购物车
 * @param {object} [extra]
 */
export function toCart(extra) {
	window.location.href = getCartSchema(extra);	
}

/**
 * 刷新购物车
 */
export function refreshCart() {
	const version = getAppVersion('jd');

	if (versionCompare(version, '5.0.0') < 0) return;

	if (isAndroid()) {
		window.AndroidNavi && window.AndroidNavi.refreshCart();
	} else if (isIOS()) {
		nativeCall('communication', { action: 'sh_refreshCart' });
	}
}

/**
 * 显示导航栏购物车按钮
 */
export function showCartButton() {
	const version = getAppVersion('jd');

	if (versionCompare(version, '5.0.0') < 0) return;

	if (isAndroid()) {
		window.AndroidNavi && window.AndroidNavi.setCart(JSON.stringify({isShow: 'Y'}));
	} else if (isIOS()) {
		if (versionCompare(version, '5.1.0') < 0) {
			nativeCall('communication', { action: 'sh_showCart' });
		} else {
			window.MobileNavi && window.MobileNavi.configBtn(JSON.stringify({
				type: 'cart', //clear_js, search, cart, custom
				display: 'show', //show, hide
				position: 'outside', //outside, inside
				icon: '',
				title: '购物车',
				jump: ''
			}));
		}
	}
}
