// @jmfe/jm-webview/src/discover.js
import { assign } from '@jmfe/jm-common';
import { buildSchema } from './call';

export function getDiscoverySchema(extra) {
	return buildSchema('virtual', assign({
		category: 'jump', 
		des: 'Discovery',
	}, extra));
}

/**
 * 进入发现频道页
 * @param {object} [extra]
 */
export function toDiscovery(extra) {
	window.location.href = getDiscoverySchema(extra);
}

export function getArticleSchema(id, extra) {
	return buildSchema('virtual', assign({
		category: 'jump', 
		des: 'faxian_article',
		id: id
	}, extra));
}

/**
 * 进入资讯详情页(发现文章)
 * @param {number} id
 * @param {object} [extra]
 */
export function toArticle(id, extra) {
	window.location.href = getArticleSchema(id, extra);
}

export function getAuthorSchema(id, extra) {
	return buildSchema('virtual', assign({
		category: 'jump', 
		des: 'faxian_author',
		authorId: id
	}, extra));
}

/**
 * 进入发布者页
 * @param {number} id
 * @param {object} [extra]
 */
export function toAuthor(id, extra) {
	window.location.href = getAuthorSchema(id, extra);
}

export function getInventorySchema(extra) {
	return buildSchema('virtual',  assign({
		category: 'jump', 
		des: 'inventory'
	}, extra));
}

/**
 * 进入发现清单
 * @param {object} [extra]
 */
export function toInventory(extra) {
	window.location.href = getInventorySchema(extra);
}

export function getInventoryDetailSchema(id, extra) {
	return buildSchema('virtual',  assign({
		category: 'jump', 
		des: 'inventoryDetail',
		id: id
	}, extra));
}

/**
 * 进入发现清单商品详情
 * @param {number} id
 * @param {object} [extra]
 */
export function toInventoryDetail(id, extra) {
	window.location.href = getInventoryDetailSchema(id, extra);
}
