// @jmfe/jm-webview/src/livecast.js
import { buildSchema } from './call';

const position = {
	find: '0',
	seckill: '2'
};

export function getLiveCastListSchema(type = 'find') {
	const des = {
		find: 'FindLivePlayList',
		seckill: 'SecKillLivePlayList'
	};
	return buildSchema('virtual', {
		category: 'jump',
		des: des[type],
		position: position[type]
	});
}

/**
 * 跳转到直播列表页
 * @param  {String} type  直播类型： find（发现）, seckill（秒杀), 默认为 find
 */
export function toLiveCastList(type = 'find') {
	window.location.href = getLiveCastListSchema(type);
}


export function getLiveCastPreviewSchema(id, type = 'find') {
	const des = {
		find: 'FindLivePreDetail',
		seckill: 'SecKillLivePreDetail'
	};
	return buildSchema('virtual', {
		category: 'jump',
		des: des[type],
		id: id,
		position: position[type]
	});
}

/**
 * 跳转到直播预告页面
 * @param  {string} id    直播 ID
 * @param  {String} type  直播类型： find（发现）, seckill（秒杀), 默认为 find
 */
export function toLiveCastPreview(id, type = 'find') {
	window.location.href = getLiveCastPreviewSchema(id, type);
}

export function getLiveCastRoomSchema(id, type = 'find') {
	return buildSchema('virtual', {
		category: 'jump',
		des: 'LivePlayerRoom',
		id: id,
		position: position[type]
	});
}

/**
 * 跳转到直播间
 * @param  {string} id    直播间 ID
 * @param  {String} type  直播类型： find（发现）, seckill（秒杀), 默认为 find
 */
export function toLiveCastRoom(id, type = 'find') {
	window.location.href = getLiveCastRoomSchema(id, type);
}
