// @jmfe/jm-webview/src/navi.js
/**
 * iOS App 文档
 * http://cf.jd.com/pages/viewpage.action?pageId=85830303
 */

import { isWebview, isApp, getAppVersion, versionCompare, isObject, isArray, isString, assign } from '@jmfe/jm-common';
/**
 * 设置webview导航背景色
 * 
 * @export
 * @param {Object} params
 * @param {string} params.naviIcon - icon颜色类型 '1'代表灰色导航栏图标（也即默认值，适用于浅色背景），'2'代表白色导航栏图标，适用于深色背景
 * @param {string} params.backgroundColor - 背景颜色, e.g.'#3a3a3a'
 * @param {string} params.pic - 背景图片url 'https://m.360buyimg.com/babel/jfs/t4036/283/2146564148/19349/f7389255/58a4184aNdb067a60.jpg'
 */
export function setNaviBackground(params) {
	const version = getAppVersion('jd');
	if (!isApp('jd') || versionCompare(version, '6.0.0') < 0) return;
	if (isWebview('wk')) {
		window.webkit.messageHandlers.MobileNavi.postMessage({
			'method': 'setNaviBackground',
			'params': JSON.stringify(params),
			'callBackName': null,
			'callBackId': null
		})
	} else {
		window.MobileNavi && MobileNavi.setNaviBackground(JSON.stringify(params));
	}
}

/**
 * 清除导航上按钮的配置，归位默认设置，在调用配置方法前使用。
 * 
 */
export function configBtnClear() {
	const version = getAppVersion('jd');
	const params = {
		type: 'clear_js',
		display: '',
		position: '',
		icon: '',
		title: '',
		jump: ''
	}
	console.log({
			'method': 'configBtnSince610',
			'params': JSON.stringify({
				clear_js: params
			}),
			'callBackName': null,
			'callBackId': null
		});
	
	if (!isApp('jd')) return;
	if (isWebview('wk')) {
		window.webkit.messageHandlers.MobileNavi.postMessage({
			method: 'configBtnSince610',
			params: JSON.stringify({
				clear_js: params
			}),
			callBackName: null,
			callBackId: null
		})
	} else {
		if (versionCompare(version, '6.1.0') < 0) {
			window.MobileNavi && MobileNavi.configBtn(JSON.stringify(params));
		} else {
			window.MobileNavi && MobileNavi.configBtnSince610(JSON.stringify({
				clear_js: params
			}));
		}
	}
}

/**
 * 5.1自定义导航栏按钮
 * 
 * @param {Object} options 
 */
export function configBtnSet(options) {
	const defaults = {
		type: '',
		display: '',
		position: '',
		icon: '',
		title: '',
		jump: '',
	};
	const type_arr = ['homepage', 'search', 'calendar', 'cart', 'message', 'custom'];

	if (!isObject(options)) return;
	if (!isApp('jd')) return;
	const params = assign(defaults, options);
	for (let i = 0; i < type_arr.length; i++) {
		if (params.type === type_arr[i]) {
			if (isWebview('wk')) {
				window.webkit.messageHandlers.MobileNavi.postMessage({
					'method': 'configBtn',
					'params': JSON.stringify(params),
					'callBackName': null,
					'callBackId': null
				})
			} else {
				window.MobileNavi && MobileNavi.configBtn(JSON.stringify(params));
			}
		}
	}
}

/**
 * 6.1自定义导航栏按钮
 * 
 * @param {Object} options 
 */
export function configBtnSince610(options) {
	let flag = '';
	const type_arr = ['homepage', 'search', 'calendar', 'cart', 'message', 'custom'];
	const _params = {};
	if (!isObject(options)) return;
	if (!isApp('jd')) return;

	for (var key in options) {
		flag = true;
		if (isObject(options[key])) {
			for (var i = 0; i < type_arr.length; i++) {
				if (options[key].type === type_arr[i]) {
					flag = false;
					break;
				}
			}
			if (!flag) {
				_params[key] = options[key];
				_params[key].display = options[key].display || '';
				_params[key].icon = options[key].icon || '';
				_params[key].jump = options[key].jump || '';
				_params[key].position = options[key].position || '';
				_params[key].title = options[key].title || '';
			}
		} else if (isArray(options[key])) {
			_params[key] = options[key];
			if (_params[key].length < 3) {
				flag = false;
				break;
			}
		} else {
			throw new Error('调用方法时传入配置对象格式错误');
		}
	}
	if (!flag) {
		if (isWebview('wk')) {
			window.webkit.messageHandlers.MobileNavi.postMessage({
				'method': 'configBtnSince610',
				'params': JSON.stringify(_params),
				'callBackName': null,
				'callBackId': null
			})
		} else {
			window.MobileNavi && window.MobileNavi.configBtnSince610(JSON.stringify(_params));
		}
	}
}

/**
 * 显示Webview左上角关闭按钮及逻辑
 * 
 */
export function showCloseBtn() {
	if (!isApp('jd')) return;
	if (isWebview('wk')) {
		window.webkit.messageHandlers.MobileNavi.postMessage({
			'method': 'showCloseBtn',
		})
	} else {
		window.MobileNavi && MobileNavi.showCloseBtn();
	}
}

/**
 * 设置透明导航栏
 * 
 * @param {Object} options 
 * @param {string} options.blackImg - 图片url https开头，与黑色icon对应
 * @param {string} options.whiteImg - 图片url https开头，与白色icon对应
 */
export function enableTransparent(options) {
	const _params = {
		blackImg: '',
		whiteImg: ''
	}
	if (!isApp('jd') || !isObject(options)) return;
	assign(_params, options);
	if (isWebview('wk')) {
		window.webkit.messageHandlers.MobileNavi.postMessage({
			'method': 'enableTransparent',
			'params': JSON.stringify(_params),
			'callBackName': null,
			'callBackId': null
		})
	} else {
		window.MobileNavi && MobileNavi.enableTransparent(JSON.stringify(_params));
	}
}

/**
 * 改变webview导航栏文字标题
 * 
 * @param {string} title - 标题文案
 */
export function setTitle(title) {
	if (!isApp('jd') || !isString(title)) return;
	if (isWebview('wk')) {
		window.webkit.messageHandlers.MobileNavi.postMessage({
			'method': 'setTitle',
			'params': title,
			'callBackName': null,
			'callBackId': null
		});
	} else {
		window.MobileNavi && MobileNavi.setTitle(title);
	}
}

/**
 * 打开扫啊扫
 * 
 * @param {string} callbackName 
 */
export function showNativeBarcodeScan(callbackName) {
	if (!isApp('jd') || !isString(callbackName)) return;
	if (isWebview('wk')) {
		window.webkit.messageHandlers.MobileNavi.postMessage({
			'method': 'showNativeBarcodeScan',
			'params': callbackName,
			'callBackName': null,
			'callBackId': null
		});
	} else {
		window.MobileNavi && MobileNavi.showNativeBarcodeScan(callbackName);
	}
}

/**
 * 多种相片上传
 * 
 */
export function showNativeUpload() {
	if (!isApp('jd')) return;
	if (isWebview('wk')) {
		window.webkit.messageHandlers.MobileNavi.postMessage({
			'method': 'showNativeUpload'
		});
	} else {
		window.MobileNavi && MobileNavi.showNativeUpload();
	}
}
