// @jmfe/jm-webview/src/other.js
/**
 * iOS App 文档
 * http://cf.jd.com/pages/viewpage.action?pageId=85830303
 */

import { isWebview, isApp, isIOS, isAndroid } from '@jmfe/jm-common';

/**
 * H5图片下载及指定目录保存
 * 
 * @param {string} [imgUrl=''] - 图片url
 * @param {string} [callBackName=''] - 全局回调函数名
 * @param {string} [callBackId=''] -  由H5自己生成的一个标示性字符串，这个是为了同时下载多张图片时h5进行区分，通过回调函数的参数返回给H5
 */
export function saveImageToPhoteAlbum(imgUrl = '', callBackName = '', callBackId = '') {
	if (!isApp('jd')) return;
	const params = { imgUrl, callBackName, callBackId }
	try {
		if (isWebview('wk')) {
			window.webkit.messageHandlers.JDAppUnite.postMessage({
				'method': 'saveImageToPhoteAlbum',
				'params': JSON.stringify(params),
				'callBackName': null,
				'callBackId': null
			})
		} else {
			window.JDAppUnite && JDAppUnite.saveImageToPhoteAlbum(JSON.stringify(params));
		}
	} catch (err) {
		throw new Error(err);
	}
}

/**
 * 微信QQ授权登录
 * 
 * @param {string} [channelName='']
 * @param {string} [callBackName='']
 */
export function bindSocialAccountWithJsonString(channelName = '', callBackName = '') {
	if (!isApp('jd')) return;
	const params = { channelName, callBackName }
	try {
		if (isWebview('wk')) {
			window.webkit.messageHandlers.MobileLogin.postMessage({
				'method': 'bindSocialAccountWithJsonString',
				'params': JSON.stringify(params),
				'callBackName': null,
				'callBackId': null,
			})
		} else {
			window.MobileLogin && MobileLogin.bindSocialAccountWithJsonString(JSON.stringify(params))
		}
	} catch (err) {
		throw new Error(err);
	}
}


/**
 * 选择或者拍摄一张图片，返回这张图片的线上url路径
 * 回调函数名 cameraCallBack  
 * 回调函数参数  e.g. jfs/t5830/165/7329426657/41247/1a8a1fcf/596f1a11N0b4984e2.jpg.webp  
 * 
 */
export function selectImage() {
	if (!isApp('jd')) return;
	if (isIOS()) {
		window.location.href = 'objc:callCamera'
	} else if (isAndroid()) {		
		window.JDClient && window.JDClient.openCamera()
	}
}

/**
 * app内关闭直播小窗
 */
export function closeLivesWindow() {
	if (!isApp('jd')) return;
	try {
		if (isWebview('wk')) {
			window.webkit.messageHandlers.JDAppUnite.postMessage({
				'method': 'closeLiveSWindow',
			})
		} else {
			window.JDAppUnite && JDAppUnite.closeLiveSWindow();
		}
	} catch (err) {
		throw new Error(err);
	}
}

/**
 * 微信签约---app内打开微信并跳到指定页面--测试
 */
export function goWxUrl(url) {
	if (!isApp('jd')) return;
	try {
		if (isWebview('wk')) {
			window.webkit.messageHandlers.JDAppUnite.postMessage({
				'method': 'signWeixinPay',
				'params': url,
			})
		} else {
			window.JDAppUnite && JDAppUnite.signWeixinPay(url);
		}
	} catch (err) {
		throw new Error(err);
	}
}

/**
 * 获取设备指纹
 */
export function getFingerInfo(cbkFunc) {
	if (!isApp('jd')) return;
	if(typeof cbkFunc != "string")throw new Error("cbkFunc need string")
	try {
		if (isWebview('wk')) {
			window.webkit.messageHandlers.JDAppUnite.postMessage({
				'method': 'getFingerInfo',
				'params' : cbkFunc
			})
		} else {
			window.JDAppUnite && JDAppUnite.getFingerInfo(cbkFunc);
		}
	} catch (err) {
		throw new Error(err);
	}
}


/**
 * 获取app导航栏高度
 */
export function getNavHeight(cbkFunc) {
	if (!isApp('jd')) return;
	if(typeof cbkFunc != "string")throw new Error("cbkFunc need string")
	try {
		if (isWebview('wk')) {
			window.webkit.messageHandlers.MobileNavi.postMessage({
				'method': 'getNaviHeight',
				'params' : cbkFunc
			})
		} else {
			window.MobileNavi && MobileNavi.getNaviHeight(cbkFunc);
		}
	} catch (err) {
		throw new Error(err);
	}
}


/**
 * webview首屏加载状态协议
 */

export function sourceIsReady() {
	if (!isApp('jd')) return;
	try {
		if (isWebview('wk')) {
			window.webkit.messageHandlers.JDAppUnite.postMessage({
				'method': 'sourceIsReady',
				'params': '1'
			})
		} else {
			window.JDAppUnite && window.JDAppUnite.sourceIsReady('1');
		}
	} catch (err) {
		throw new Error(err);
	}
}
