// @jmfe/jm-webview/src/share.js
import { assign, isAndroid, isIOS } from '@jmfe/jm-common';
import { nativeCall } from './call';

const shareData = {
	title: '',
	content: '',
	shareUrl: '',
	iconUrl: '',
	qrparam: null,
	callback: null,
	clickcallback: null
};

/**
 * 呼出原生分享菜单
 * @param {object} opts
 */
export function callSharePane(opts) {
	if (isAndroid()) {
		nativeCall('communication', assign({
			action: 'share',
			type: '111'
		}, shareData, opts));
	} else if (isIOS()) {
		nativeCall('communication', assign({
			action: 'share'
		}, shareData, opts));
	}
}

/**
 * 设置分享内容
 * @param {object} opts
 */
export function setShareInfo(opts) {
	let data = assign({}, shareData, opts);
	
	if (isAndroid()) {
		if (shareHelper) {
			shareHelper.setShareInfo(opts.title, opts.content, opts.shareUrl, opts.iconUrl);
		} else {
			nativeCall('communication', assign({
				action: 'share',
				type: '111'
			}, shareData, opts));			
		}
	} else if (isIOS()) {
		nativeCall('communication', assign({
			action: 'syncShareData'
		}, shareData, opts));
	}
}

export function sendDirectShare() {
	
}
