// @jmfe/jm-common/src/type.js
/**
 * 对象类型相关判断
 * @module type
 */

/**
 * 获取指定对象的类型
 * @param  {type} obj 	对象
 * @return {string}     对象类型值
 */
function getType(obj) {
	return Object.prototype.toString.call(obj);
}

/**
 * 判断是否是function
 * @param  {object}  obj 对象
 * @return {Boolean}
 */
export function isFunction(obj) {
	return typeof obj === 'function';
}

/**
 * 判断是否是Object
 * @param  {object}  obj 对象
 * @return {Boolean}
 */
export function isObject(obj) {
	return '[object Object]' === getType(obj);
}

/**
 * 判断是否是array
 * @param  {object}  obj 对象
 * @return {Boolean}
 */
export function isArray(obj) {
	return '[object Array]' == getType(obj);
}


/**
 * 判断是否是undefined
 * @param  {object}  obj 对象
 * @return {Boolean}
 */
export function isUndefined(obj) {
	return '[object Undefined]' === getType(obj);
}

/**
 * 判断是否是null
 * @param  {object}  obj 对象
 * @return {Boolean}
 */
export function isNull(obj) {
	return '[object Null]' == getType(obj);
}


/**
 * 判断是否是字符串
 * @param  {object}  obj 对象
 * @return {Boolean}
 */
export function isString(obj) {
	return typeof obj === 'string';
}

/**
 * 判断是否是布尔类型
 * @param  {object}  obj 对象
 * @return {Boolean}
 */
export function isBoolean(obj) {
	return typeof obj === 'boolean';
}

/**
 * 判断是否是一个数字
 * @param  {number}  num 数字
 * @return {Boolean}
 */
export function isNumber(num) {
	return '[object Number]' === getType(num);
}
